File: 백준/Silver/29728. 실버와 소수는 둘다 S로 시작한다/실버와 소수는 둘다 S로 시작한다.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int n;
  static int sCount, bCount;
  static char front, back;
  static final int LIMIT = 5000000;

  public static void main(String[] args) throws IOException {
    setUp();

    boolean[] primeSieve = getPrimeSieve();
    for (int i = 1; i < n + 1; i++) {
      if (!primeSieve[i]) {
        bCount++;
        back = 'B';
      } else if (back == 'B') {
        bCount--;
        sCount += 2;
        back = 'S';
        swap(i);
      } else {
        sCount++;
        back = 'S';
        swap(i);
      }

      if (i == 1) {
        front = back;
      }
    }

    sb.append(bCount).append(" ").append(sCount);

    output();
  }

  private static void swap(int index) {
    if (index == 1 || index == 2) {
      front = back;
      return;
    }
    char tmp = front;
    front = back;
    back = tmp;
  }

  private static boolean[] getPrimeSieve() {
    boolean[] sieve = new boolean[LIMIT + 1];
    Arrays.fill(sieve, true);
    sieve[1] = false;

    for (int i = 2; i * i < LIMIT + 1; i++) {
      if (sieve[i]) {
        for (int j = i * i; j < LIMIT + 1; j += i) {
          sieve[j] = false;
        }
      }
    }

    return sieve;
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    n = Integer.parseInt(br.readLine());
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}

File: 백준/Silver/29728. 실버와 소수는 둘다 S로 시작한다/실버와 소수는 둘다 S로 시작한다.py
import sys
from collections import deque

input = sys.stdin.readline

LIMIT = 5000000


def get_prime_sieve():
    sieve = [True] * (n + 1)
    sieve[1] = False

    for i in range(2, int(n**0.5) + 1):
        if sieve[i]:
            for j in range(i * i, n + 1, i):
                sieve[j] = False

    return sieve


n = int(input())

prime_sieve = get_prime_sieve()
dq = deque()
is_back = True

for curr in range(1, n + 1):
    if not prime_sieve[curr]:
        if is_back:
            dq.append("B")
        else:
            dq.appendleft("B")
    elif is_back and dq[-1] == "B":
        dq.pop()
        dq.append("S")
        dq.append("S")
        is_back = not is_back
    elif not is_back and dq[0] == "B":
        dq.popleft()
        dq.appendleft("S")
        dq.appendleft("S")
        is_back = not is_back
    elif is_back:
        dq.append("S")
        is_back = not is_back
    else:
        dq.appendleft("S")
        is_back = not is_back


b_count, s_count = 0, 0
while dq:
    curr = dq.pop()
    if curr == "B":
        b_count += 1
    else:
        s_count += 1

print(b_count, s_count)